### PR TITLE
feat: import config metadata and import to `projectSettings`

### PR DESCRIPTION
**Pull request by tomasciccola**
tied to https://github.com/digidem/comapeo-mobile/issues/527

it uses an unpublished version of mapeo-schema with `configName` and `configVersion` in `projectSettings`

I think this PR should be the last to be merged, after updating `@mapeo/schema`, `@mapeo/mock-data`, `@mapeo/default-config`

### CODE REVIEW

File: src/config-import.js [hunk 1]
```diff
@@ -270,6 +281,27 @@ async function findTranslationsFile(entries) {
...
+/**
+ * @param {ReadonlyArray<Entry>} entries
+ * @returns {Promise<MetadataFile | undefined>}
+ */
+async function findMetadataFile(entries) {
+  const metadataEntry = entries.find(
+    (entry) => entry.filename === 'metadata.json'
+  )
+  if (!metadataEntry) return
+  //assert(metadataEntry, 'Zip file does not contain metadata.json')
```
Comment (tomasciccola) [inline src/config-import.js:293]: since metadata, translations and presets files are loaded in parallel (see line 52), failing where it can't find a `metadata.json` file messes assertions in tests.

File: src/config-import.js [hunk 2]
```diff
@@ -479,6 +511,24 @@ function parseIcon(filename, buf) {
   }
 }
 
+/**
+ * @param {Record<string,unknown>} obj
+ * @returns {obj is MetadataFile}
+ */
+function isValidMetadataFile(obj) {
```
Comment (tomasciccola) [inline src/config-import.js:518]: should we allow extra fields in `metadata.json` I would say since it would mean not needing to update the config builder, and we can ignore values we don't want

**Comment by EvanHahn** [comment]
Let me know when I should take a look at this (it's currently a draft). Happy to review whenever!

**Comment by tomasciccola** [comment]
> Let me know when I should take a look at this (it's currently a draft). Happy to review whenever!

I think there's some doubts I should clear out before moving forward (see stuff asked here -> https://github.com/digidem/comapeo-mobile/issues/527). 
Most of the work is done, just need to fix tests and type errors...

File: src/config-import.js (hunk 2 above)
Comment (EvanHahn) [reply src/config-import.js:521]: Yes, I think we should allow extra fields. This seems good to me.

File: tests/fixtures/config/completeConfig.zip
Comment (EvanHahn) [inline tests/fixtures/config/completeConfig.zip:1]: _Not_ for this PR, but it'd be nice to be able to see the contents of these in source without having to unzip them manually.

File: tests/config-import.js [hunk 3]
```diff
@@ -83,9 +123,8 @@ test('config import - icons', async () => {
   let config = await readConfig(
     './tests/fixtures/config/invalidIconFilename.zip'
   )
-  /* eslint-disable-next-line */
-  for await (const icon of config.icons()) {
-  }
+
+  await arrayFrom(config.icons())
```
Comment (EvanHahn) [inline tests/config-import.js:127]: Minor: I'd move these to a separate PR, as it's not strictly related. But I don't feel strongly.

**Review by EvanHahn — COMMENTED** [review]
Draft looks good so far!

File: tests/fixtures/config/completeConfig.zip
Comment (tomasciccola) [reply tests/fixtures/config/completeConfig.zip:1]: Yes, that IS a drag. Don't even mention about updating each zip file where data changes. Probably gonna address it in the future (or now in a separate PR). I think having the unzipped files and _maybe_ zip them as a sort of `pre-test` step?

**Comment by EvanHahn** [comment]
Lemme know when I should take a look at this.

File: tests/fixtures/config/completeConfig.zip
Comment (EvanHahn) [reply tests/fixtures/config/completeConfig.zip:1]: That's exactly what I'd do. Have them unzipped in the source code and zip them when it's time to run the test. But again, not for this PR.

File: src/config-import.js [hunk 4]
```diff
@@ -36,23 +39,31 @@ const MAX_ICON_SIZE = 10_000_000
...
+    `invalid or missing config file version ${metadataFile.fileVersion}. We support version ${SUPPORTED_CONFIG_VERSION}}`
+  )
 
   return {
     get warnings() {
       return warnings
     },
 
+    get metadata() {
+      return { ...metadataFile, importDate: new Date().toISOString() }
```
Comment (gmaclennan) [inline src/config-import.js:64]: this will return a different value every time this property is accessed. The date should be set when import happens.

File: src/config-import.js [hunk 5]
```diff
@@ -36,23 +39,31 @@ const MAX_ICON_SIZE = 10_000_000
...
     throw new Error(`Zip file contains too many entries. Max is ${MAX_ENTRIES}`)
   }
   const entries = await zip.readEntries(MAX_ENTRIES)
-  const [presetsFile, translationsFile] = await Promise.all([
-    findPresetsFile(entries),
-    findTranslationsFile(entries),
-  ])
+  const presetsFile = await findPresetsFile(entries)
+  const translationsFile = await findTranslationsFile(entries)
+  const metadataFile = await findMetadataFile(entries)
```
Comment (gmaclennan) [inline src/config-import.js:52]: why switch to serial vs. parallel?

File: src/config-import.js (hunk 2 above)
Comment (gmaclennan) [reply src/config-import.js:519]: How about using ajv and JSONSchema for this? We already have them as deps via mapeo-schema. Could be more work than necessary for now.

Comment (tomasciccola) [reply src/config-import.js:527]: (heh, was already doing that: MetadataFile is grabbed from @mapeo/schema)

**Comment by EvanHahn** [comment]
I'm catching up after some time off. I'll plan to review this tomorrow or Wednesday.

File: tests/fixtures/config/completeConfig.zip
Comment (EvanHahn) [inline tests/fixtures/config/completeConfig.zip:1]: Probably makes sense to merge #739, then rebase and update these ZIPs.

**Review by EvanHahn — APPROVED** [review]
LGTM but we probably want to merge #739 first.

**Comment by gmaclennan** [comment]
Just a note for myself more than anything to say I did a quick review and all looks good.
